packages/packit/packit-api.nix

```diff
@@ -34,6 +34,7 @@ let
     nativeBuildInputs = [ perl ];
     installPhase = ''
       find $GRADLE_USER_HOME/caches/modules-2 -type f -regex '.*\.\(jar\|pom\|module\)' \
+         | LC_ALL=C sort \
```

Comment:
not just this line but i have no clue what youre doing here, this is definitely bside the point of this PR but could you leave a comment about this or something just briefly explaining this

Reply:
Ahah yes, this line is dark magic stolen from occurrences I found in `nixpkgs`. See for example 
https://github.com/NixOS/nixpkgs/blob/8121f3559a98259a8e767dedf4eaf3939442c54d/pkgs/applications/file-managers/mucommander/default.nix#L39-L48


So Nix builds are either sandboxes with no network, or they have network access but need to have an exact hash output. If we wanted to build a package that fetches external dependencies using stuff gradle, cargo, npm, we'd have to set the hash of the output of the build, which is tedious since that would check any time our source changes just a tiny bit, or we change the build process a little.

The compromise in Nix is to split the build process in two, fetch the dependencies out of the sandbox and set a fixed hash, and then run the actual build in the sandbox and not have to set the output hash. This is fine for sensible package managers, but gradle is not one of them. It doesn't really have a proper way of just fetching the dependencies.

What we do here is build packit-api twice. After the first build, we throw away the build output but keep the cache in `$GRADLE_USER_HOME/caches/modules-2`, and do some preprocessing to make the cache have the same file layout as Maven does. We set a hash for this (`sources.gradleDepsHash`). Then on the second build we replace the maven references with the output of the first build (it's what the `gradleInit` below does).

Now the actual sort I added is because the same artifact might appear multiple times in the cache, with slightly different contents, eg. one was published to maven central and one to the gradle plugin repository. Absolute madness but here we are. Thanfully from what I could tell, the differences were very minor.

We can only have a single copy in our fake maven repo. `find` was acting non deterministically, and the order wasn't stable, and as a result we weren't always copying the same file. I was getting different result on my machine compared to CI. The sort helps keep that deterministic.

`LC_ALL=C` is slang for "set the locale to naive English". Sorting technically depends on the configure language, so that help keep the result obvious. It probably doesn't matter, since everything is ASCII and I doubt Nix lets the host system locale leak through.

I hate all of this, but it's the best I could find for now. nixpkgs has actually improved this quite a bit in the current master, but we need to wait until November for a stable release: https://github.com/NixOS/nixpkgs/pull/272380

Reply:
that was actually fucked up